snuba/consumers/strict_consumer.py

```diff
@@ -0,0 +1,141 @@
+            'group.id': group_id,
+            'enable.partition.eof': 'true',
+            'default.topic.config': {
+                'auto.offset.reset': auto_offset_reset,
+            },
+        }
+
+        self.__consumer = self.create_consumer(consumer_config)
+
+        def on_partitions_assigned(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("New partitions assigned: %r", partitions)
+            self.__on_partitions_assigned(consumer, partitions)
+
+        def on_partitions_revoked(
+            consumer: Consumer,
+            partitions: Sequence[TopicPartition],
+        ):
+            logger.info("Partitions revoked: %r", partitions)
+            self.__on_partitions_revoked(consumer, partitions)
+
+        self.__consumer.subscribe(
+            [topic],
+            on_assign=on_partitions_assigned,
+            on_revoke=on_partitions_revoked,
+        )
+
+    def create_consumer(self, config) -> Consumer:
+        return Consumer(config)
+
+    def run(self) -> None:
+        watermarks = {}
+
+        message = self.__consumer.poll(timeout=self._assign_timeout)
+        if not message:
+            # Since we enabled enable.partition.eof, no message means I could not
+            # consume and it is different from the case I managed to consume and
+            # I reached the end of the partition. Thus we fail.
+            raise NoPartitionAssigned("No partition was assigned within timeout")
+
+        while not self.__shutdown:
+            error = message.error()
+            if error:
+                if error.code() == KafkaError._PARTITION_EOF:
+                    logger.debug("End of topic reached")
+                    self.__consumer.close()
+                    return
+                else:
+                    raise Exception(message.error())
+
+            commit_decision = self.__on_message(message)
+            if commit_decision == CommitDecision.COMMIT_THIS:
+                self.__consumer.commit()
+            elif commit_decision == CommitDecision.COMMIT_PREV:
+                prev_watermark = watermarks.get(
+                    (message.partition(), message.topic()),
+                )
+                if prev_watermark is not None:
+                    commit_pos = TopicPartition(
+                        partition=message.partition(),
+                        topic=message.topic(),
+                        offset=prev_watermark,
+                    )
+                    self.__consumer.commit(commit_pos)
+                else:
+                    logger.debug(
+                        "No previous message to commit on partition %s on topic %s",
+                        message.partition(),
+                        message.topic(),
+                    )
+
+            watermarks[(message.partition(), message.topic())] = message.offset()
+            message = self.__consumer.poll(timeout=1.0)
```

Comment:
Should this be called if we're shutting down?

Reply:
The result should be the same, no?

Reply:
Yeah, ultimately it could just have to wait longer. I guess it depends on what happens with the change to partition EOF, this is likely not even reachable on the last loop iteration since it'll exit early if there's nothing to fetch (each loop should either return a message or EOF, and never actually wait for a significant amount of time, I think?)

Reply:
Actually I think we should not run it when shutting down.
We can shut down in two ways:
- EOF. In this case there is nothing to poll when shutting down. We actually do not want to consume further if any new message comes in.
- a shutdown signal. In this case it does not matter. It means we did not consume till the end. Every message consumed is supposed to leave the topic in a state we can recover from upon restart.